File: f1tenth_waypoint/src/waypoint_generator_node.cpp
#include "f1tenth_waypoint/waypoint_generator_node.hpp"

Waypoint::Waypoint(const std::string &name) : Node(name)
{

    // Set up the waypoint type and topic
    std::tie(message_type, waypoint_topic) = this->setup_waypoint_option();

    // Create subscriptions based on the message type
    if (!message_type.empty() && !waypoint_topic.empty())
    {
        if (message_type == "PointStamped")
        {
            // Initialize the subscription for PointStamped
            clicked_point_sub_ = this->create_subscription<geometry_msgs::msg::PointStamped>(
                waypoint_topic, 10, std::bind(&Waypoint::waypoint_callback, this, std::placeholders::_1));
        }
        else if (message_type == "Odometry")
        {
            // Initialize the subscription for Odometry
            odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
                waypoint_topic, 10, std::bind(&Waypoint::odom_callback, this, std::placeholders::_1));
        }
    }

    // Optionally, set up a publisher for markers
    waypoint_marker_pub_ = this->create_publisher<visualization_msgs::msg::Marker>("/f1tenth_waypoint_marker", 10);
}

void Waypoint::initialize()
{
        // Initialize ClickedWaypoint and odom_waypoint_ with the current node instance

    // odom_waypoint_ = std::make_shared<OdomWaypoint>(this->shared_from_this());
    // clicked_waypoint_ = std::make_shared<ClickedWaypoint>(this->shared_from_this());


    if (message_type == "Odometry")
    {
        // Only initialize OdomWaypoint
        odom_waypoint_ = std::make_shared<OdomWaypoint>(this->shared_from_this());
        RCLCPP_INFO(this->get_logger(), "Initialized OdomWaypoint");
    }
    else if (message_type == "PointStamped")
    {
        // Only initialize ClickedWaypoint
        clicked_waypoint_ = std::make_shared<ClickedWaypoint>(this->shared_from_this());
        RCLCPP_INFO(this->get_logger(), "Initialized ClickedWaypoint");
    }
}

std::pair<std::string, std::string> Waypoint::setup_waypoint_option()
{
    std::string odom_topic;
    std::string clicked_point_topic;
    std::string waypoint_option;

    try
    {
            // Declare the parameters

        if (!this->has_parameter("odom_topic"))
        {
            this->declare_parameter<std::string>("odom_topic", "/odom");
        }

        if (!this->has_parameter("clicked_point_topic"))
        {
            this->declare_parameter<std::string>("clicked_point_topic", "/clicked_point");
        }

        if (!this->has_parameter("waypoint_options"))
        {
            this->declare_parameter<std::string>("waypoint_options", "clicked_waypoint"); // clicked_waypoint   odom_waypoint
        }
        
        // Get the parameters
        odom_topic = this->get_parameter("odom_topic").as_string();
        clicked_point_topic = this->get_parameter("clicked_point_topic").as_string();
        waypoint_option = this->get_parameter("waypoint_options").as_string();

    }
    catch(const rclcpp::exceptions::ParameterNotDeclaredException &e) 
    {
        RCLCPP_WARN(this-> get_logger(), "Could not get parameters ... setting variables to default");
        RCLCPP_WARN(this-> get_logger(), "Error: %s", e.what());
    }
    catch(const std::exception &e)
    {
        RCLCPP_WARN(this->get_logger(), "An error occured: %s", e.what());
        
    }
    



    // Set the message type and topic based on the waypoint type
    if (waypoint_option == this->CLICKED_WAYPOINT)
    {
        RCLCPP_INFO(this->get_logger(), "Using CLICKEDPOINT_WAYPOINT option");

        return {"PointStamped", clicked_point_topic};
    }
    else if (waypoint_option == this->ODOM_WAYPOINT)
    {
        RCLCPP_INFO(this->get_logger(), "Using ODOM_WAYPOINT option");
        return {"Odometry", odom_topic};
    }
    else
    {
        RCLCPP_INFO(this->get_logger(), "Unknown waypoint option, using default");
        return {"", ""};
    }
}

 //rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr rvizPoint_pub_;

void Waypoint::waypoint_callback(const geometry_msgs::msg::PointStamped::SharedPtr waypoint_msgs)
{
    if (waypoint_msgs != nullptr)
    {
        RCLCPP_INFO(this->get_logger(), "Received PointStamped message: x = %.2f, y = %.2f, z = %.2f",
                    waypoint_msgs->point.x, waypoint_msgs->point.y, waypoint_msgs->point.z);

        clicked_waypoint_->clicked_point_callback(waypoint_msgs, waypoint_marker_pub_);
    }
}

void Waypoint::odom_callback(const nav_msgs::msg::Odometry::SharedPtr odom_msgs)
{
    if (odom_msgs != nullptr)
    {
        RCLCPP_INFO(get_logger(), "Received Odometry message: x = %.2f, y = %.2f",
                    odom_msgs->pose.pose.position.x, odom_msgs->pose.pose.position.y);
        odom_waypoint_->odom_callback(odom_msgs, waypoint_marker_pub_);
    }
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto waypoint_node = std::make_shared<Waypoint>("waypoint_generator_node");
    waypoint_node->initialize();
    rclcpp::spin(waypoint_node);
    rclcpp::shutdown();
    return 0;
}

File: f1tenth_waypoint/f1tenth_waypoint_generator/waypoint_generator_node.py
#!/usr/bin/env python3
import sys
from ament_index_python import get_package_prefix 
site_packages_path = f'{get_package_prefix("f1tenth_waypoint")}/lib/python3.8/site-packages'
if site_packages_path not in sys.path: sys.path.append(site_packages_path)


import rclpy
from rclpy.node import Node
from enum import Enum
from visualization_msgs.msg import Marker
from typing import  Union

from nav_msgs.msg import Odometry
from f1tenth_waypoint_generator.waypoint_generation_options.record_clicked_waypoints import ClickedWaypoint
from f1tenth_waypoint_generator.waypoint_generation_options.record_odom_waypoints import OdomWaypoint
from geometry_msgs.msg import PointStamped, Point



class WaypointType(Enum):
    CLICKEDPOINT_WAYPOINT = "clicked_waypoint"
    ODOM_WAYPOINT = "odom_waypoint"

class WaypointGenerator(Node):
    def __init__(self, waypoint_type:WaypointType, odom_waypoint:OdomWaypoint, clicked_waypoint:ClickedWaypoint)->None:
        super().__init__('waypoint_generator_node')

        self.odom_waypoint = odom_waypoint
        self.clicked_waypoint = clicked_waypoint

        message_type, waypoint_topic = self.setup_waypoint_option(waypoint_type)

        if message_type is not None and waypoint_topic is not None:
            self.clicked_point_sub_ = self.create_subscription(message_type, waypoint_topic, self.waypoint_callback,10)
        
        self.waypoint_marker_pub_ = self.create_publisher(Marker, '/f1tenth_waypoint_marker', 10)
    
    def setup_waypoint_option(self, waypoint_type:WaypointType):
        
        if not self.has_parameter('waypoint_options'):
            self.declare_parameter(name='waypoint_options', value='clicked_waypoint')  # clicked_waypoint, odom_waypoint
        
        if not self.has_parameter('clicked_point_topic'):
            self.declare_parameter(name='clicked_point_topic', value='/clicked_point')
        
        if not self.has_parameter('odom_topic'):
            self.declare_parameter(name='odom_topic', value='/odom')
            
        try:
            odom_topic = self.get_parameter('odom_topic').get_parameter_value().string_value
            self.get_logger().info(f'\t odom_topic:\t "{odom_topic}')

            clicked_point_topic = self.get_parameter('clicked_point_topic').get_parameter_value().string_value
            self.get_logger().info(f'\t clicked_point_topic:\t "{clicked_point_topic}')

            waypoint_option = self.get_parameter('waypoint_options').get_parameter_value().string_value
            self.get_logger().info(f'\t waypoint_options:\t "{waypoint_option}')

        except Exception as e:
            self.get_logger().warn('Could not get parameters ...setting variables to default')
            self.get_logger().warn(f'Error: "{e}"')
            return None, None

        if waypoint_option == waypoint_type.CLICKEDPOINT_WAYPOINT.value:
            self.get_logger().info("Using CLICKEDPOINT_WAYPOINT option")
            return PointStamped, clicked_point_topic

        elif waypoint_option == waypoint_type.ODOM_WAYPOINT.value:
            self.get_logger().info("Using ODOM_WAYPOINT option")
            return Odometry, odom_topic
        else:
            self.get_logger().info("Unknown waypoint option, using default")
            return None,None
    
    def waypoint_callback(self, waypoint_msgs: Union[PointStamped, Odometry])->None:
        """
        Callback function to handle incoming waypoint message. Can handle either PointStamped or Odometry message
        """
        # Check if the message is of type PointStamped
        if isinstance(waypoint_msgs, PointStamped):
            self.get_logger().info(f"Received PointStamped message: x = {waypoint_msgs.point.x}, y = {waypoint_msgs.point.y}, z = {waypoint_msgs.point.z}")
            self.clicked_waypoint.clicked_point_callback(waypoint_msgs, self.waypoint_marker_pub_ )
            

        # Check if the message is of type Odometry
        elif isinstance(waypoint_msgs, Odometry):
            # self.get_logger().info(f"Received Odometry message: x = {waypoint_msgs.pose.pose.position.x}, y = {waypoint_msgs.pose.pose.position.y}, z = {waypoint_msgs.pose.pose.position.z}")
            self.odom_waypoint.odom_callback(waypoint_msgs, self.waypoint_marker_pub_)

        else:
            self.get_logger().warn(f"Unknown message type received: {type(waypoint_msgs)}")


def main(args=None):
    rclpy.init(args=args)

    share_node = rclpy.create_node('waypoint_shared_node')
    clicked_waypoint = ClickedWaypoint(share_node); odom_waypoint = OdomWaypoint(share_node)
    
    waypoint_node = WaypointGenerator(WaypointType, odom_waypoint, clicked_waypoint)
        
    try:
        rclpy.spin(waypoint_node)
    except KeyboardInterrupt:
        pass
    
    waypoint_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
